PhysicsTools/PatAlgos/src/SoftMuonMvaEstimator.cc

```diff
@@ -77,18 +77,70 @@ void SoftMuonMvaEstimator::computeMva(const pat::Muon& muon)
 	
 	iValidFraction_ =  iTrack->validFraction();
 	innerChi2_ =  iTrack->normalizedChi2();
-	layersWithMeasurement_ = static_cast<float>(iTrack->hitPattern().trackerLayersWithMeasurement());
+	layersWithMeasurement_ = iTrack->hitPattern().trackerLayersWithMeasurement();
 
 	outerChi2_ = oTrack->normalizedChi2();
 
-	Qprod_ =  static_cast<float>(iTrack->charge()) * static_cast<float>(oTrack->charge());
+	qProd_ =  iTrack->charge()*oTrack->charge();
 
-	int vDT, vCSCmax, vRPC;
-	vDT = gTrack->hitPattern().numberOfValidMuonDTHits();
-	vCSCmax = gTrack->hitPattern().numberOfValidMuonCSCHits();
-	if(vCSCmax>6) vCSCmax=6;
-	vRPC = gTrack->hitPattern().numberOfValidMuonRPCHits();
-	vMuonHitComb_ = (vDT/2 + vCSCmax + vRPC);
+	const reco::HitPattern &gMpattern = gTrack->hitPattern();
+
+	std::vector<int> fvDThits = {0,0,0,0};
+	std::vector<int> fvRPChits = {0,0,0,0};
+	std::vector<int> fvCSChits = {0,0,0,0};
+
+	vMuonHitComb_ = 0;
+
+	for (int i=0;i<gMpattern.numberOfAllHits(reco::HitPattern::TRACK_HITS);i++){
+
+	  uint32_t hit = gMpattern.getHitPattern(reco::HitPattern::TRACK_HITS, i);
+	  if ( !gMpattern.validHitFilter(hit) ) continue;
+
+	  if (gMpattern.getMuonStation(hit) == 1){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[0]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[0]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[0]++;
+	  }
+  	
+	  if (gMpattern.getMuonStation(hit) == 2){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[1]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[1]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[1]++;
+	  }
+
+	  if (gMpattern.getMuonStation(hit) == 3){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[2]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[2]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[2]++;
+	  }
+
+	  if (gMpattern.getMuonStation(hit) == 4){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[3]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[3]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[3]++;
+	  }
+
+	}
+  
+
+	if (!((fvDThits.size() < 4)
+	   || (fvRPChits.size() < 4)
+	   || (fvCSChits.size() < 4))){
```

Comment:
each vector is declared with size 4 
``` std::vector<int> fvDThits = {0,0,0,0};```
and there is no code modifying the size of the vector
what's the point of this check ``` (fvDThits.size() < 4) ```? it's always false.

Reply:
You are right, this is a copy/paste mistake from my side. I will fix it.